### PR TITLE
[RFC]: shell auto test framework for sof.

### PR DESCRIPTION
**Pull request by stevyan**
# Shell Automation Test
## Feature List
Shell automation test (sat) framework is a shell script-based, extensible keyword-driven automation framework for testing on the Linux environment. It supports following features:
- test target project via terminal command with vary options;
- specify the test project, and create the test project if not exist; [TODO], currently, use fixed test project (sof-test-suites) as default;
- support run one or multiple test suites for once executing via option `-s "<suite_a>;<suite_b>;<suite...>"`;
- support create the test suite via command line if not exist; [TODO]
- support flexible extensions; currently only implement the test result listening extension, include 3 parts (project | suite | case) & cooresponding pharses (prepare | processing | finished[passed/failed/blocked]), it is easy for collecting the test result for output, e.g.: csv, html or anything else; Extension mechanism is implemented via defined the virtual function tables ("_\_vlist\__")
- support the log collection during the whole test procedure, and can dump the running environment log data;
- support the flexible configuration for different platform or product by changing config file (*_suite.conf*) [TODO]

## Structure
- __at-fw__: automation test framework source code;
- __Test Project__: a project which need be tested via sat, all relative cases need be written with shell scripts;
- __Test Suite__: a set of cases which can decribed the same proposal, e.g.: functionality-test, base-test, sanity-test, and so forth;
- __Test Case__: executable shell script for testing a function, there is a template of case (_case.t) in the _template_ folder, the user can copy the *_case.t* as based scripts, and only need fill the corresponding code to relative block;

## Specific files for SAT
- __*\_init\_*__ : project or suite initial files, the corresponding environment initialization or preparation can be written here; the executing route likes that:
    >  ` tproject/_init_ ->  tproject/suite_a/_init_ -> tproject/suite_b/_init_ -> tproject/suite_b/subfolder/_init_`
- __*\_uninit\_*__ : project or suite uninitialized files, when run suite/project completed, the corresponding environment need be clean or dealed with here; the executing route likes that:
    >   `tproject/suite_b/subfolder/_uninit_ -> tproject/suite_b/_uninit_ ->  tproject/suite_a/_init_ -> tproject/_init_`
- __*\_suite.conf*__ : the project/suite relatived config file which provides some options for project/suite executing; If the config file isn't present, the parent config file will be used; (currently, only implement base fixed configuration file for the sof-test-suites project, need be improved to more flexible configuration in near furture.)[TBC]

- __*.tscl*__ : the test case list for a suite/sub-folder, if *.tscl* used, the framework will go through whole list file, and execute the ordered cases one by one strictly; otherwise, the all cases of current folder will be executed randomly;

- __*\_\_keywords\_\_*__ : the keyword definitions for the corresponding test project, any keyword, which want to be used in the project scope, should be defined as a export function, e.g.:
    ```
    function foo
    {
        echo do something
    }

    declare -fx foo
    ```
- __*\_vlist\_*__: the virtual function list for extension, currently, only test result extension has implemented, more extension can be added as demanded with the same mechanism;

## Usage
```
usage: sat [options]
Run the automation test suites according specified test project.

  -p {PROJECT}                  Specified the test project which contains all relative test suites for this project;
                                if -p isn't present, current default project will be used. (sof-test-suites),
                                if the {PROJECT} doesn't exist, it will prompt user to confirm if need create a {PROJECT};
                                if -f option present, the {PROJECT} will be created automatically.

  -s {SUITE[;SUITE2;...]}       Select test suite {SUITE}, if not exist, create one manually or automatically (-f present);

  -o {REPORT_ROOT}              Test report output directory, default output is under test project, named 'report' (TODO);
  -C {KEY=VAL[;KEY2=VAL2;...]}  Specified the customized definition varibles which need saved as project level;
                                e.g.: tplg=sof-apl-pcm512x.tplg;key=val;

  -c {KEY=VAL[;KEY2=VAL2;...]}  Specified the customized definition varibles which need saved to specified test suites,
                                it will check the option -s as target test suites;
                                e.g.: -s platform/byt/mnb -c tplg=sof-apl-pcm512x.tplg;key=val;

  -e {SUTIE=y|n[;SUITE2=y|n;...]}  Enable/disable the test suite with following format: 
                                   test_suite1=y;test_suite2=n;test_suite3/sub_suite=y; the suite is the relative path of project root;


  -f Force confirmed when the new project or case suite missed, and need create automatically;
  -u Update the whole project's status if cases/suites changed. (TODO)
  -l List the project's test suites or suites' cases (TODO);
  -x Execute the test project or test suites (need match the option -s for specified executing test suites);
  -h Show this help;

example:
    1. run the suite1 of myProject
    sat -p myProject -s suite1 -x

    2. query how many test suites in the myProject
    sat -p myProject -l

    3. add the new test suites in the myProject
    sat -p myProject -s newSuite

    4. set the common varible for myProject
    sat -p myProject -C MY_VAR=value1

    5. set the common varible for suite2 of myProject
    sat -p myProject -s suite2 -c SUB_VAR=value2


satwrapper is a simple wrapper of sof test project with sat, and it will auto detect the test platform for sof:
usage: satwrapper tplg_file

    tplg_file   Specified the topology file which need be tested.
```

## Example
`./sat -C tplg=sof-apl-nocodec.tplg -x -s base-test`
>[INFO]: === CASE PASSED ===  
>[INFO]: END| {5 - verify_pcm_list.sh}: Executing Done!  
>[INFO]: BEG| {6 - check_sof_logger.sh}: Executing Case...  
>rm: cannot remove '/tmp/sof-logger.log': Operation not permitted  
>[INFO]: Case passed  
>  adding: check_sof_logger.sh.txt (stored 0%)  
>[INFO]: === CASE PASSED ===  
>[INFO]: END| {6 - check_sof_logger.sh}: Executing Done!  
>[INFO]: BEG| {7 - check_aplay_function.sh}: Executing Case...  
>aplay: main:788: audio open error: Invalid argument  
>[INFO]: Case FAILED via PCM: Port0  
>aplay: main:788: audio open error: Invalid argument  
>[INFO]: Case FAILED via PCM: Port1  
>aplay: main:788: audio open error: Invalid argument  
>[INFO]: Case FAILED via PCM: Port3  
>aplay: main:788: audio open error: Invalid argument  
>[INFO]: Case FAILED via PCM: Port5  
>[INFO]: Skip Device [hw:0,6](capture)...  
>[INFO]: Ignore Case via PCM: DMIC01  
>[INFO]: Case Failed  
>  adding: manifest.xml (deflated 67%)  
>  adding: sof-apl.ldc (deflated 86%)  
>  adding: sof-sys-dbg/ (stored 0%)  
>  adding: sof-sys-dbg/fw_version (deflated 27%)  
>  adding: sof-sys-dbg/inbox (deflated 98%)  
>  adding: sof-sys-dbg/pp (deflated 97%)  
>  adding: sof-sys-dbg/trace (deflated 70%)  
>  adding: sof-sys-dbg/regs (deflated 97%)  
>  adding: sof-sys-dbg/debug (deflated 97%)  
>  adding: sof-sys-dbg/exception (deflated 96%)  
>  adding: sof-sys-dbg/hda (deflated 99%)  
>  adding: sof-sys-dbg/stream (deflated 98%)  
>  adding: sof-sys-dbg/outbox (deflated 98%)  
>  adding: sof-sys-dbg/etrace (deflated 98%)  
>  adding: sof-sys-dbg/dsp (deflated 100%)  
>  adding: dmesg.txt (deflated 88%)  
>  adding: sof-logger (deflated 62%)  
>  adding: check_aplay_function.sh.txt (deflated 66%)  
>[INFO]: === CASE FAILED ===  
>[INFO]: END| {7 - check_aplay_function.sh}: Executing Done!  
>[INFO]: End executing test suite: [base-test]  
>[INFO]: ----------------------------------------------------------------  
>[REPORT]: ========================== Summary ==============================  
>[REPORT]: Test Date: 03/07/19 16:37:28  
>[REPORT]: Case Suites: [/home/ubuntu/sofcitest/./sof-test-suites/base-test]  
>[REPORT]: Total Cases : 7  
>[REPORT]: Passed Count: 5  
>[REPORT]: Failed Count: 2  
>[REPORT]: Pass Rate: 71%  
>[REPORT]: ================================================================  
>[INFO]: => zip the suite log file [/home/ubuntu/sofcitest/./out/base-test/result.txt] ...  
>  adding: result.txt  
>	zip warning:  file size changed while zipping /home/ubuntu/sofcitest/./out/base-test/result.txt  
> (deflated 79%)  
>init shell plugin... _v8mywkUM_spg_uninit  
>init shell plugin... _ejBlNVdu_spg_uninit  
>[INFO]: uninitialize the [tcr] plugin.

### CODE REVIEW

File: tools/test/sofcitest/LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2019 sof-ci
```
Comment (lgirdwood) [inline tools/test/sofcitest/LICENSE]: Why is this MIT licence ?. The Copyright needs to be Intel. Is this your own work or derived from somewhere else ?

**Comment by lgirdwood** [comment]
Can you provide more detail on how this can test topology, and what it tests in the topology. We also are missing copyright and licence. Is this PR importing an existing testing framework ?

**Comment by stevyan** [comment]
@lgirdwood topology will be passed as a input parameters with option "-C tplg=<tplg_name>",  <tplg_name> can be a short file name or full file path, if short file name,  it will attempt to find the topology binary from "/lib/firmware/intel/sof-tplg/" by default. 
Regarding to test topology, the __tplgreader__ will list all pipelines which are stored at specified topology binary, and test cases will go through those pipeline as demand, e.g.: test aplay, test loopback, and so on.

For license issue, this testing framework doesn't derive from anyway, it is a new one, but the invoked __tplgreader__ (generated by tplgreader.c) tool perhaps re-use some alsa topology relative code, @xiulipan, could you please help declare this MIT license issue?

**Comment by xiulipan** [comment]
@stevyan @lgirdwood 
Please remove the license. This file is added when create by the system when setting up the the local dev git repo, we did not need that.
And add header for each files to make sure there is no license issue.

**Comment by xiulipan** [comment]
@lgirdwood 
Do you think the test should go as a standalone project? As tests are actually run on the DUT but sof git repo is always used on some DEV machine.

**Comment by lgirdwood** [comment]
@xiulipan this has a tight coupling with the topology header so best to include under the main SOF repo to keep everything aligned..

**Comment by sofci** [comment]
Can one of the admins verify this patch?

**Comment by lgirdwood** [comment]
@stevyan any update here ?

**Comment by tlauda** [comment]
@stevyan Ping.

**Comment by stevyan** [comment]
Following sat features are ongoing:
- support running this framework both on Host PC/Server and DUT device;
- support execute the single case via option '-x <case_name>' ;
- support multi-level configure files (_.rconfig_ - running time config, _sconfig_ - preset shell config)

I'll update the commit after those features completed. Expected target date: ww20.05

File: tools/test/sofcitest/sat-fw/at-plugin [hunk 2]
```diff
@@ -0,0 +1,102 @@
+#!/bin/bash
```
Comment (jajanusz) [inline tools/test/sofcitest/sat-fw/at-plugin]: Please add spdx

**Comment by tlauda** [comment]
@stevyan Any update?

**Comment by wenqingfu** [comment]
@DarrenxWang @stripes416 @fredoh9 @xiulipan 

Steve doesn't work on this any more.

**Comment by wenqingfu** [comment]
should this PR be closed for now? I think it's obsolete now. @xiulipan

**Pull request closed without merging**